### PR TITLE
add junit test for stash file parameter

### PR DESCRIPTION
**Pull request by olamy**

### CODE REVIEW

**Comment by olamy** [comment]
@jglick please have a look there is a bit of testing and documentation. 
Maybe we can merge that and have a first release? please let me know.
Thanks

File: pom.xml [hunk 1]
```diff
@@ -30,6 +30,11 @@
                 <scope>import</scope>
                 <type>pom</type>
             </dependency>
+            <dependency>
```
Comment (oleg-nenashev) [inline pom.xml:33]: https://github.com/jenkinsci/jackson2-api-plugin can be used instead

**Review by oleg-nenashev — APPROVED** [review]

File: pom.xml (hunk 1 above)
Comment (olamy) [reply pom.xml]: good catch. fixed

File: README.md [hunk 2]
```diff
@@ -12,20 +12,45 @@ Offers alternative types of file parameter that are compatible with Pipeline and
 
 See [JENKINS-27413](https://issues.jenkins-ci.org/browse/JENKINS-27413) and [JENKINS-29289](https://issues.jenkins-ci.org/browse/JENKINS-29289) for background.
 
+## Usage in declarative pipeline
+
+You can now declare file parameters as is in declarative pipeline:
+
+```
```
Comment (jglick) [inline README.md:19]:
```suggestion
```groovy
```

File: README.md [hunk 3]
```diff
@@ -12,20 +12,45 @@ Offers alternative types of file parameter that are compatible with Pipeline and
...
+pipeline {
+  agent any
+  parameters {
+    base64File(name: 'FILE')
+    stashed64File(name: 'FILE-STASH')
```
Comment (jglick) [inline README.md:24]: Does
```suggestion
    base64File 'FILE'
    stashed64File 'FILE-STASH'
```

work?

File: README.md [hunk 4]
```diff
@@ -12,20 +12,45 @@ Offers alternative types of file parameter that are compatible with Pipeline and
...
+  }
+  stages {
+    stage('Example') {
+      steps {
+        withFileParameter('FILE') {
+          echo(/loaded '${readFile(FILE)}' from $FILE/)                                                    
+        }
+        unstash "FILE-STASH"
+        echo(/loaded '${readFile("./FILE-STASH")}'/)
```
Comment (jglick) [inline README.md:33]:
```suggestion
        unstash 'FILE-STASH'
        echo(/loaded '${readFile('FILE-STASH')}'/)
```

File: pom.xml [hunk 5]
```diff
@@ -42,6 +42,11 @@
             <groupId>org.jenkins-ci.plugins.workflow</groupId>
             <artifactId>workflow-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>jackson2-api</artifactId>
+            <scope>test</scope>
+        </dependency>
```
Comment (jglick) [inline pom.xml:49]: Why is this here?

File: src/main/java/io/jenkins/plugins/file_parameters/Base64FileParameterDefinition.java [hunk 6]
```diff
@@ -54,6 +54,12 @@
             return "Base64 File Parameter";
         }
 
+        @Override public String getHelpFile()
+        {
```
Comment (jglick) [inline src/main/java/io/jenkins/plugins/file_parameters/Base64FileParameterDefinition.java:58]:
```suggestion
        @Override public String getHelpFile() {
```

File: src/main/java/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition.java [hunk 7]
```diff
@@ -54,6 +54,11 @@
             return "Stashed File Parameter";
         }
 
+        @Override public String getHelpFile()
```
Comment (jglick) [inline src/main/java/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition.java:57]: You should never need to override `getHelpFile` except when you want to share one file among multiple descriptors. Instead, simply create `help.html` in the same directory as `config.jelly`.

File: src/main/webapp/help-base64File.html [hunk 8]
```diff
@@ -0,0 +1,20 @@
+<div>
+  <p>File parameter compatible with Pipeline, how to use it in a declarative pipeline:</p>
+  <pre>
```
Comment (jglick) [inline src/main/webapp/help-base64File.html:3]: probably need `<code>` also

File: src/main/webapp/help-base64File.html [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  agent any
+  parameters {
+    base64File(name: 'THEFILE')
+  }
+  stages {
+    stage('Example') {
+      steps {
+        withFileParameter('THEFILE') {
+        echo(/loaded '${readFile(FILE)}' from $FILE/)
+        }
```
Comment (jglick) [inline src/main/webapp/help-base64File.html:14]: check indentation / nesting

File: README.md [hunk 10]
```diff
@@ -12,20 +12,45 @@ Offers alternative types of file parameter that are compatible with Pipeline and
...
+  agent any
+  parameters {
+    base64File(name: 'FILE')
+    stashed64File(name: 'FILE-STASH')
+  }
+  stages {
+    stage('Example') {
+      steps {
+        withFileParameter('FILE') {
+          echo(/loaded '${readFile(FILE)}' from $FILE/)
```
Comment (jglick) [inline README.md:30]: Probably more intuitive to say
```suggestion
          sh 'cat $FILE'
```

File: README.md [hunk 11]
```diff
@@ -12,20 +12,45 @@ Offers alternative types of file parameter that are compatible with Pipeline and
...
+pipeline {
+  agent any
+  parameters {
+    base64File(name: 'FILE')
+    stashed64File(name: 'FILE-STASH')
+  }
+  stages {
+    stage('Example') {
+      steps {
+        withFileParameter('FILE') {
```
Comment (jglick) [inline README.md:29]: IIRC Declarative supports using block-scoped steps in an `option` block. Does it work here?

File: src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java [hunk 12]
```diff
@@ -58,4 +62,129 @@
...
+            "        }\n" +
+            "      }\n" +
+            "    }\n" +
+            "  }\n" +
+            "}";
+
+        p.setDefinition(new CpsFlowDefinition(pipeline, true));
+        WorkflowRun run = p.scheduleBuild2(0).get();
+        r.waitForCompletion(run);
+        // definitely will fail but we just ensure parameter has been created
```
Comment (jglick) [inline src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java:117]: You could use `[WorkflowJob.]addProperty` to define params before running the build. Or the build may be able to check at runtime

```groovy
env.FILE != null
```

(For Base64 params. Not for stashed params, which never set an env var at the build level.)

File: src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java [hunk 13]
```diff
@@ -58,4 +62,129 @@
...
+    }
+
+    @Test public void stashedDeclarativeParameterCreated() throws Exception {
+        r.createSlave("remote", null, null);
+        WorkflowJob p = r.createProject(WorkflowJob.class, "myjob");
+
+        String pipeline = "pipeline {\n" +
+            "    agent any\n" +
+            "    parameters {\n" +
+            "        stashed64File(name: 'FILE-STASH')\n" +
```
Comment (jglick) [inline src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java:158]: I think I messed up the symbol, should be
```suggestion
            "        stashedBase64File(name: 'FILE-STASH')\n" +
```

**Review by jglick — COMMENTED** [review]
Stuff to work with here. I would want to do some fixups before merging.

As to a release, I think it is prudent to look into `input` / `build` first. Also there is an unimplemented method. Need to look over the list of outstanding items, implement the easy ones, and refile the rest as GH issues.

File: pom.xml (hunk 5 above)
Comment (olamy) [reply pom.xml:49]: if not we got this 
```
[WARNING] Rule 5: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message:
Failed while enforcing RequireUpperBoundDeps. The error(s) are [
Require upper bound dependencies error for com.fasterxml.jackson.datatype:jackson-datatype-json-org:2.9.8 paths to dependency are:
+-io.jenkins.plugins:file-parameters:1.0-SNAPSHOT
  +-org.jenkinsci.plugins:pipeline-model-definition:1.7.2
    +-org.jenkinsci.plugins:pipeline-model-api:1.7.2
      +-com.fasterxml.jackson.datatype:jackson-datatype-json-org:2.9.8
and
+-io.jenkins.plugins:file-parameters:1.0-SNAPSHOT
  +-org.jenkinsci.plugins:pipeline-model-definition:1.7.2
    +-org.jenkinsci.plugins:pipeline-model-api:1.7.2
      +-org.jenkins-ci.plugins:jackson2-api:2.12.1 (managed) <-- org.jenkins-ci.plugins:jackson2-api:2.9.8
        +-com.fasterxml.jackson.datatype:jackson-datatype-json-org:2.12.1
, 
Require upper bound dependencies error for org.json:json:20171018 paths to dependency are:
+-io.jenkins.plugins:file-parameters:1.0-SNAPSHOT
  +-org.jenkinsci.plugins:pipeline-model-definition:1.7.2
    +-org.jenkinsci.plugins:pipeline-model-api:1.7.2
      +-com.fasterxml.jackson.datatype:jackson-datatype-json-org:2.9.8
        +-org.json:json:20171018
and
+-io.jenkins.plugins:file-parameters:1.0-SNAPSHOT
  +-org.jenkinsci.plugins:pipeline-model-definition:1.7.2
    +-org.jenkins-ci.plugins:git-client:3.6.0
      +-org.jenkins-ci.plugins:trilead-api:1.0.13
        +-org.jenkins-ci:trilead-ssh2:build-217-jenkins-27
          +-com.google.crypto.tink:tink:1.5.0
            +-org.json:json:20180813
```

File: src/main/java/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition.java (hunk 7 above)
Comment (olamy) [reply src/main/java/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition.java:57]: there is no config.jelly for this as it extends AbstractFileParameterDefinition

File: src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java (hunk 12 above)
Comment (olamy) [reply src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java:117]: the idea here was test the auto creation of the parameter after the first run

File: README.md (hunk 11 above)
Comment (olamy) [reply README.md:29]: yep it works

File: src/main/webapp/help-base64File.html (hunk 8 above)
Comment (olamy) [reply src/main/webapp/help-base64File.html:3]: `<code>`  doesn't have a good rendering for this case
Screen Shot 2021-01-23 at 2 22 41 pm

**Comment by olamy** [comment]
@jglick looks good to me to look into `build`/`input`. we can merge this PR then I can take care of it.

File: pom.xml (hunk 5 above)
Comment (jglick) [reply pom.xml:49]: Hmm @timja sounds like a `bom` failure. Sound familiar? `sample-plugin/pom.xml` includes this dep, which it ought not.

File: src/main/java/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition.java (hunk 7 above)
Comment (jglick) [reply src/main/java/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition.java:57]: You put it in the directory `config.jelly` would have been in: `src/main/resources/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition/`

File: src/main/webapp/help-base64File.html (hunk 8 above)
Comment (jglick) [reply src/main/webapp/help-base64File.html:3]: I meant `<pre>` _and_ `<code>`. If `<pre>` alone triggers a monospaced font then I guess that suffices.

File: README.md (hunk 11 above)
Comment (jglick) [reply README.md:29]: But you prefer to demonstrate using a block-scoped step inside `steps`?

File: src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java [hunk 14]
```diff
@@ -58,4 +62,129 @@
...
 
+    @Ignore("need to implement option to tolerate undefined parameter")
+    @Test public void base64Undefined() throws Exception {
+        r.createSlave("remote", null, null);
+        WorkflowJob p = r.createProject(WorkflowJob.class, "myjob");
+        p.addProperty(new ParametersDefinitionProperty(new Base64FileParameterDefinition("FILE", null)));
+        String pipeline = "pipeline {\n" +
+            "  agent any\n" +
+            "  parameters {\n" +
+            "    base64File(name: 'FILE')\n" +
```
Comment (jglick) [inline src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java:73]: You are not testing the simplified syntax? How do you know it works?

**Review by jglick — DISMISSED** [review]

File: pom.xml (hunk 5 above)
Comment (timja) [reply pom.xml:49]: are you saying that `sample-plugin` shouldn't have jackson2-api @jglick ?

File: src/main/java/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition.java (hunk 7 above)
Comment (olamy) [reply src/main/java/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition.java:57]: 🤦‍♂️

File: pom.xml (hunk 5 above)
Comment (jglick) [reply pom.xml:49]: Well it should not _need_ it, and it looks like the fact that it does have that explicit dependency is masking a problem.

File: src/main/webapp/help-base64File.html (hunk 8 above)
Comment (timja) [reply src/main/webapp/help-base64File.html]: normally you would use pre and code together...

File: src/main/java/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition.java [hunk 15]
```diff
@@ -47,16 +47,11 @@
 
     // TODO equals/hashCode
 
-    @Symbol("stashed64File")
+    @Symbol("stashedBase64File")
```
Comment (jglick) [inline src/main/java/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition.java:50]: Oh what am I thinking. I meant
```suggestion
    @Symbol("stashedFile")
```

File: src/main/java/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition.java [hunk 16]
```diff
@@ -47,16 +47,11 @@
...
     @Extension public static final class DescriptorImpl extends ParameterDefinition.ParameterDescriptor {
 
         @Override public String getDisplayName() {
-            return "Stashed File Parameter";
-        }
-
-        @Override public String getHelpFile()
-        {
-            return "/plugin/file-parameters/help-stashedFile.html";
+            return "Stashed Base64 File Parameter";
```
Comment (jglick) [inline src/main/java/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition.java:54]:
```suggestion
            return "Stashed File Parameter";
```

File: README.md [hunk 17]
```diff
@@ -21,7 +21,7 @@ pipeline {
   agent any
   parameters {
     base64File 'FILE'
-    stashed64File 'FILE-STASH'
+    stashedBase64File 'FILE-STASH'
```
Comment (jglick) [inline README.md:24]:
```suggestion
    stashedFile 'FILE-STASH'
```

File: src/main/resources/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition/help.html [hunk 18]
```diff
@@ -4,7 +4,7 @@
 pipeline {
   agent any
   parameters {
-    stashed64File(name: 'FILE-STASH')
+    stashedBase64File(name: 'FILE-STASH')
```
Comment (jglick) [inline src/main/resources/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition/help.html:7]:
```suggestion
    stashedFile(name: 'FILE-STASH')
```

or, if it works,
```suggestion
    stashedFile 'FILE-STASH'
```

File: src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java [hunk 19]
```diff
@@ -155,7 +155,7 @@
         String pipeline = "pipeline {\n" +
             "    agent any\n" +
             "    parameters {\n" +
-            "        stashed64File(name: 'FILE-STASH')\n" +
+            "        stashedBase64File 'FILE-STASH'\n" +
```
Comment (jglick) [inline src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java:158]:
```suggestion
            "        stashedFile 'FILE-STASH'\n" +
```

**Review by jglick — COMMENTED** [review]
Sorry I must have been tired!

File: src/main/resources/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition/help.html (hunk 18 above)
Comment (olamy) [reply src/main/resources/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition/help.html:7]: simple syntax not working

File: README.md (hunk 11 above)
Comment (olamy) [reply README.md:29]: uhm I missed that one but I'm not sure what sort of syntax you mean?

File: src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java (hunk 14 above)
Comment (olamy) [reply src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java]: simplified syntax doesn't work

File: pom.xml (hunk 5 above)
Comment (olamy) [reply pom.xml:49]: I guess it will be fixed when this change https://github.com/jenkinsci/pipeline-model-definition-plugin/commit/2c186f56aa74557c77b5f1be0d6dcd045680eaef will be release and then new version integrated in the bom
but until all of this done we need this dependency (as `test`) or this:
```
        <dependency>
            <groupId>org.jenkinsci.plugins</groupId>
            <artifactId>pipeline-model-definition</artifactId>
            <scope>test</scope>
            <exclusions>
                <exclusion>
                    <groupId>com.fasterxml.jackson.datatype</groupId>
                    <artifactId>jackson-datatype-json-org</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
```
personally I do not have any strong opinion on what workaround to use as they are bot `test` dependencies.

File: README.md [hunk 20]
```diff
@@ -12,20 +12,45 @@ Offers alternative types of file parameter that are compatible with Pipeline and
...
+  - [X] inline help text
   - [X] `withFileParameter` compatibility
   - [X] manual test
   - [X] automated test
 - [ ] stashed file parameter (suitable for larger files used in Pipeline)
   - [X] implementation
-  - [ ] inline help text
-  - [ ] `withFileParameter` compatibility
+  - [X] inline help text
+  - [X] `withFileParameter` compatibility (manual test)
```
Comment (jglick) [inline README.md:51]: I do not think it could possibly work currently: https://github.com/jenkinsci/file-parameters-plugin/blob/b2e08aa74bb2d655c92a989c2fa191dd6fc16cf4/src/main/java/io/jenkins/plugins/file_parameters/StashedFileParameterValue.java#L83

File: src/main/java/io/jenkins/plugins/file_parameters/Base64FileParameterDefinition.java [hunk 21]
```diff
@@ -53,7 +53,8 @@
         @Override public String getDisplayName() {
             return "Base64 File Parameter";
         }
-
     }
 
+
+
```
Comment (jglick) [inline src/main/java/io/jenkins/plugins/file_parameters/Base64FileParameterDefinition.java:59]: could be reverted

File: src/main/resources/io/jenkins/plugins/file_parameters/Base64FileParameterDefinition/help.html [hunk 22]
```diff
@@ -0,0 +1,41 @@
+<div>
+  <p>File parameter compatible with Pipeline, how to use it in a declarative pipeline:</p>
```
Comment (jglick) [inline src/main/resources/io/jenkins/plugins/file_parameters/Base64FileParameterDefinition/help.html:2]: The critical thing to mention here is that this parameter type is only suitable for small files.

File: src/main/resources/io/jenkins/plugins/file_parameters/Base64FileParameterDefinition/help.html [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+pipeline {
+  agent any
+  parameters {
+    base64File(name: 'THEFILE')
+  }
+  stages {
+    stage('Example') {
+      steps {
+        withFileParameter('THEFILE') {
+          echo(/loaded '${readFile(FILE)}' from $FILE/)
```
Comment (jglick) [inline src/main/resources/io/jenkins/plugins/file_parameters/Base64FileParameterDefinition/help.html:13]:
```suggestion
          sh 'cat $FILE'
```

File: src/main/resources/io/jenkins/plugins/file_parameters/Base64FileParameterDefinition/help.html [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+  <pre>
+pipeline {
+  agent any
+  parameters {
+    base64File(name: 'THEFILE')
+  }
+  stages {
+    stage('Example') {
+      steps {
+        withFileParameter('THEFILE') {
```
Comment (jglick) [inline src/main/resources/io/jenkins/plugins/file_parameters/Base64FileParameterDefinition/help.html:12]: Or you can simply

```groovy
sh 'echo $FILE | base64 -d | whatever'
```

File: src/main/resources/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition/help.html [hunk 25]
```diff
@@ -0,0 +1,19 @@
+<div>
+  <p>File parameter compatible with Pipeline but this one stash the file, how to use it in a declarative pipeline:</p>
```
Comment (jglick) [inline src/main/resources/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition/help.html:2]: Should explain that the point of this parameter type is to handle big files.

File: src/main/resources/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition/help.html [hunk 26]
```diff
@@ -0,0 +1,19 @@
...
+pipeline {
+  agent any
+  parameters {
+    stashedFile(name: 'FILE-STASH')
+  }
+  stages {
+    stage('Example') {
+      steps {
+        unstash "FILE-STASH"
+        echo(/loaded '${readFile("./FILE-STASH")}'/)
```
Comment (jglick) [inline src/main/resources/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition/help.html:13]:
```suggestion
        sh 'cat FILE-STASH'
```

**Review by jglick — COMMENTED** [review]
Sorry the back-and-forth across time zones is a little crazy!

File: README.md (hunk 20 above)
Comment (olamy) [reply README.md:51]: well not sure using such usage is very useful :) but true we can be error prone...

```
pipeline {
    agent any
    parameters {
        stashedFile(name: 'FILE')
    }
    stages {
        stage('Example') {
            steps {
                withFileParameter(name:'FILE') {
                    unstash "FILE"
                    sh 'cat $FILE'
                }
            }
        }
    }
}

**Comment by olamy** [comment]
> Sorry the back-and-forth across time zones is a little crazy!

no worries especially with adding weekend, public holidays etc... 😄

File: README.md (hunk 20 above)
Comment (jglick) [reply README.md:51]: If you are using `withFileParameter` then you are not supposed to need `unstash` (and vice-versa).

Comment (jglick) [reply README.md:51]:
```suggestion
  - [ ] `withFileParameter` compatibility
```

File: src/main/resources/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition/help.html [hunk 27]
```diff
@@ -0,0 +1,19 @@
+<div>
+  <p>File parameter compatible with Pipeline but this one stash the file and must be used to handle bug files, how to use it in a declarative pipeline:</p>
```
Comment (jglick) [inline src/main/resources/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition/help.html:2]:
```suggestion
  <p>File parameter compatible with Pipeline but this one stash the file and must be used to handle big files, how to use it in a declarative pipeline:</p>
```

File: src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java [hunk 28]
```diff
@@ -58,4 +62,153 @@
...
+        String pipeline = "pipeline {\n" +
+            "  agent any\n" +
+            "  parameters {\n" +
+            "    base64File (name: 'FILE')\n" +
+            "  }\n" +
+            "  stages {\n" +
+            "    stage('Example') {\n" +
+            "      steps {\n" +
+            "        withFileParameter('FILE') {\n" +
+            "          echo(/loaded '${readFile(FILE).toUpperCase(Locale.ROOT)}' from $FILE/) \n" +
```
Comment (jglick) [inline src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java:131]: BTW while `sh 'cat $FILE'` is most realistic for docs, I prefer `readFile` for functional tests because it requires no special work to pass on Windows.

File: src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java [hunk 29]
```diff
@@ -58,4 +62,153 @@
...
+        assertThat(new CLICommandInvoker(r, "build").
+                       withStdin(new ByteArrayInputStream("uploaded content here".getBytes())).
+                       invokeWithArgs("-f", "-p", "FILE=", "myjob"),
+                   CLICommandInvoker.Matcher.succeeded());
+        WorkflowRun b = p.getBuildByNumber(2);
+        assertNotNull(b);
+        r.assertLogContains("loaded 'UPLOADED CONTENT HERE'", b);
+    }
+
+    @Test public void stashed() throws Exception {
```
Comment (jglick) [inline src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java:158]: Fine but it is not using `FileParameterWrapper` so does not belong here. (Unless the two are made to be compatible and the test made to exercise that.)

File: src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java [hunk 30]
```diff
@@ -58,4 +62,153 @@
...
+        assertThat(new CLICommandInvoker(r, "build").
+                       withStdin(new ByteArrayInputStream("uploaded content here".getBytes())).
+                       invokeWithArgs("-f", "-p", "FILE-STASH=", "myjob"),
+                   CLICommandInvoker.Matcher.succeeded());
+        WorkflowRun b = p.getBuildByNumber(1);
+        assertNotNull(b);
+        r.assertLogContains("loaded 'UPLOADED CONTENT HERE'", b);
+    }
+
+    @Test public void stashedDeclarativeParameterCreated() throws Exception {
```
Comment (jglick) [inline src/test/java/io/jenkins/plugins/file_parameters/FileParameterWrapperTest.java:175]: Ditto.

**Review by jglick — COMMENTED** [review]
Enough to merge, but I have various follow-ups and corrections.

File: src/main/resources/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition/help.html (hunk 18 above)
Comment (jglick) [reply src/main/resources/io/jenkins/plugins/file_parameters/StashedFileParameterDefinition/help.html]: addressed in another PR

File: README.md (hunk 11 above)
Comment (jglick) [reply README.md]: Bug in Declarative it seems: https://issues.jenkins.io/browse/JENKINS-64761
